internal/promotion/runner/builtin/http_requester.go

```diff
@@ -235,16 +233,22 @@ func (h *httpRequester) wasRequestSuccessful(
 	case cfg.SuccessExpression != "":
 		program, err := expr.Compile(cfg.SuccessExpression)
 		if err != nil {
-			return false, fmt.Errorf("error compiling success expression: %w", err)
+			return false, &promotion.TerminalError{
+				Err: fmt.Errorf("error compiling success expression %q: %w", cfg.SuccessExpression, err),
+			}
 		}
 		successAny, err := expr.Run(program, env)
 		if err != nil {
-			return false, fmt.Errorf("error evaluating success expression: %w", err)
+			return false, &promotion.TerminalError{
+				Err: fmt.Errorf("error evaluating success expression %q: %w", cfg.SuccessExpression, err),
+			}
```

Comment:
If the failure evaluating the expression was attributable to applying the expression to an unexpected response, it's possible that evaluation of the expression might succeed on retry. I think that possibility may also be applicable to the expression evaluating as a non-bool (below). Also applicable in the logic of the `didRequestFail()` method.

Reply:
Fixed!